tests/unit/test_token_response.py

```diff
@@ -0,0 +1,236 @@
+        Confirms only refresh_token and token_type may be missing,
+        And expires_at is created correctly.
+        """
+        top_convert = _convert_token_info_dict(self.top_token)
+
+        # expected results for top token data
+        for key in ["scope", "access_token", "refresh_token", "token_type"]:
+            self.assertEqual(self.top_token[key], top_convert[key])
+
+        # sanity check expires at, assuming test will run well within 1 second
+        self.assertEqual(int(time.time()) + self.top_token["expires_in"],
+                         top_convert["expires_at_seconds"])
+
+        # missing refresh_token or token_type
+        for key in ["refresh_token", "token_type"]:
+            value = self.top_token.pop(key)
+            top_convert = _convert_token_info_dict(self.top_token)
+            self.assertIsNone(top_convert[key])
+            self.top_token[key] = value
+
+        # required keys
+        for key in ["scope", "access_token"]:
+            value = self.top_token.pop(key)
+            with self.assertRaises(KeyError):
+                _convert_token_info_dict(self.top_token)
+            self.top_token[key] = value
+
+        # no expires_in makes expires at now
+        self.top_token.pop("expires_in")
+        top_convert = _convert_token_info_dict(self.top_token)
+        self.assertEqual(int(time.time()), top_convert["expires_at_seconds"])
+
+    def test_by_resource_server(self):
+        """
+        Gets by_resource_server attribute from test response,
+        Confirms expected values found for top and other tokens
+        """
+        by_server = self.response.by_resource_server
+
+        # confirm data by server matches known token values
+        for server, token in [("server1", self.top_token),
+                              ("server2", self.other_token1),
+                              ("server3", self.other_token2)]:
+            server_data = by_server[server]
+            for key in ["scope", "access_token",
+                        "refresh_token", "token_type"]:
+                self.assertEqual(server_data[key], token[key])
+            # assumes test will run well within 1 second
+            self.assertEqual(int(time.time()) + token["expires_in"],
+                             server_data["expires_at_seconds"])
+
+    def test_expires_at_seconds(self):
+        """
+        Gets response's expires_at_seconds attribute, confirms expected value
+        """
+        self.assertEqual(self.response.expires_at_seconds,
+                         int(time.time()) + self.top_token["expires_in"])
+
+    def test_expires_in(self):
+        """
+        Gets two measurements of response's expires_in attribute,
+        Sanity checks that expires_in decreases as time goes on
+        """
+        first_measure = self.response.expires_in
+        time.sleep(1)
+        second_measure = self.response.expires_in
+        self.assertTrue(first_measure < self.top_token["expires_in"])
```

Comment:
Maybe I've missed something. Why should this be true? Wouldn't you expect them to be equal most of the time, and rarely different?
If this is right, please add a comment explaining.